ql/src/java/org/apache/hadoop/hive/ql/stats/ColStatsProcessor.java

```diff
@@ -220,8 +220,10 @@ public int persistColumnStats(Hive db, Table tbl) throws HiveException, MetaExce
       start = System. currentTimeMillis();
       if (tbl != null && tbl.isNonNative() && tbl.getStorageHandler().canSetColStatistics(tbl)) {
         tbl.getStorageHandler().setColStatistics(tbl, colStats);
+      } else {
+        // Set table or partition column statistics in metastore.
+        db.setPartitionColumnStatistics(request);
       }
-      db.setPartitionColumnStatistics(request);
```

Comment:
I'm a little confused about the change.
If we can not get stats from puffine due to some exception, we can fallback get stats from metastore. So i think maybe write stats into the two places is meaningful.  Please correct me if i misunderstand. Thansk.

Reply:
@zhangbutao I agree with your point. However, storing stats in 2 places has its pros & cons - 
Pros - 
1. We can fallback to metastore by changing the config - `hive.iceberg.stats.source=metastore` if we are able to not able to get stats from Puffin files.

Cons - 
1. Any change in Puffin files by external clients is not visible to metastore.
2. Performance effect of executing these metastore DB calls to store column stats.

In the approach mentioned in the PR, if users want to use metastore to get stats if they are not able to get stats from Puffin, then set `hive.iceberg.stats.source=metastore` and execute `ANALYZE TABLE <tableName> COMPUTE STATISTICS FOR COLUMNS`. (This will have an overhead of one more ANALYZE query).

I will leave it to the community to decide if its best to store stats in 2 places or storing it in a single place is sufficient. If the community thinks that this it is best to store in 2 places, then I won't proceed further. Otherwise, I will continue with the patch.

Reply:
> If we can not get stats from puffine due to some exception, we can fallback get stats from metastore. So i think maybe write stats into the two places is meaningful

Storing at two places have additional costs during write & currently we have two modes, "iceberg" & "metastore", so both denotes where to store the stats.

Storing at both sides, seems to be a third mode, like "both" and presently we don't have a fallback logic either during read side, that if puffin file are inaccessible then go to metastore kind of thing. 

May be if we want such a thing, we can have a new mode, if we feel that is required in future stages.

As of now, I think, "iceberg" mode should store only in puffin and "metastore" mode should store only in "metastore"

Reply:
https://github.com/apache/hive/blob/10600392ec05bd50351b4668b38dd84502a7eb72/ql/src/java/org/apache/hadoop/hive/ql/optimizer/StatsOptimizer.java#L308-L311  

https://github.com/apache/hive/blob/10600392ec05bd50351b4668b38dd84502a7eb72/ql/src/java/org/apache/hadoop/hive/ql/optimizer/StatsOptimizer.java#L932-L950
 Currently, Like this example HIVE-27347 always uses the iceberg basic stats from metatstore to optimize` count(*)` query. We should consider how to do this if only using puffin stats.

Reply:
created https://github.com/apache/hive/pull/5400. to address above